### PR TITLE
Add kelp to cool and temperate shallow ocean

### PR DESCRIPTION
**Pull request by paramat**
Add kelp to cool and temperate shallow ocean

Original texture by tobyplowy, recoloured by paramat.
////////////

screenshot_20171010_210836

Addresses #1129 

Appears with varying density in cool and temperate shallow ocean (not icesheet, tundra, rainforest, savanna or hot desert).
Height is random between 3 and 6 nodes. Base cube depth is from y = -5 to y = -10, so that the top never passes above the water surface.
The whole plant takes the light of the base cube, so i have limited it to a depth of -10 so that the plant does not become very dark. Real Kelp requires light anyway.
Noie is used so that the kelp forests cover one thrd of the suitable areas. Not rare but not too common. As we add more underwater plants in future it will become less barren.

I used tobyplowy's texture which is good, but i have colourised it to a kelp yellow-green, tobyplowy i hope you are ok with this, also, is CC BY-SA 3.0 ok as a license? This is the standard MTG texture license.

### CODE REVIEW

**Comment by jastevenson303** [comment]
I use opaque water and am dreading testing this, but will do so immediately.  This will make a fantastic addition to coral.

**Comment by paramat** [comment]
Yes i noticed the inventory image is a little strange, i think it's extruded then rotated to isometric view or something.

screenshot_20171006_011445

Added an 'on place' function to the node. On placement a random height is chosen and the top node is checked to be water source, then protection is checked for both top and base nodes.
As with waterlily, 'on rightclick' for the pointed node is checked.

**Comment by jastevenson303** [comment]
can be altered by specifying 'inventory_image' should make it 2d

**Comment by paramat** [comment]
Thanks will do.

**Comment by Fixer-007** [comment]
Will be nice to have more than one type of those sea things.

**Comment by paramat** [comment]
We have corals too, but certainly we could have more than 2, another PR. Something 1 node high would be good.

**Comment by paramat** [comment]
Updated to the first attempt at a tuned distribution, here's an example of distribution and typical maximum density:

screenshot_20171007_231049

screenshot_20171007_231447

To aviod too much rendering load i have made kelp forest cover one third of seafloor area, but within those areas the density gets satisfyingly high.

**Comment by tobyplowy** [comment]
@paramat Yes of course I'm ok with that why would i not be :)
Yes CC BY-SA 3.0 is good

**Comment by tobyplowy** [comment]
@paramat also sorry for the long wait :/

**Comment by Ezhh** [comment]
Yes yes yes yes yes. Looks great.

Would like a way to vary height as mentioned above, but this is just the type of sea content I want to see in game. :)

**Comment by paramat** [comment]
Random param2 for decorations added to TODO list.

**Comment by Ezhh** [comment]
Any chance of more than one colour of these?

**Comment by paramat** [comment]
Maybe but not sure.

WIP again will add engine param2 randomisation first.

**Comment by Ezhh** [comment]
Different colours actually make a huge difference. I suspect I'll just add something in a mod somewhere. :)

**Comment by paramat** [comment]
Yeah i guess it does make a significant difference.

**Comment by paramat** [comment]
This needs additional engine changes to be usable with the new random param2 feature https://github.com/minetest/minetest/pull/6512
Currently i am using a 1-node schematic for kelp as this is the only way to have the sand base placed in the seabed instead of on it. However the random param2 feature cannot affect schematic nodes.

I will need to place kelp as simple decorations but add engine code to 'y-offset' the placement. I recently added a 'y-offset' parameter for schematic decorations, i could make this usable with simple decorations too.

**Comment by paramat** [comment]
https://github.com/minetest/minetest/pull/6512 merged.

**Comment by paramat** [comment]
https://github.com/minetest/minetest/pull/6514 merged.

File: mods/default/nodes.lua [hunk 1]
```diff
@@ -1449,6 +1451,51 @@ minetest.register_node("default:acacia_bush_sapling", {
...
+	special_tiles = {{name = "default_kelp.png", tileable_vertical = true}},
+	inventory_image = "default_kelp.png",
+	paramtype2 = "leveled",
+	groups = {snappy = 3},
+	node_placement_prediction = "",
+
+	on_place = function(itemstack, placer, pointed_thing)
+		local node_ptu = minetest.get_node(pointed_thing.under)
+		local def_ptu = minetest.registered_nodes[node_ptu.name]
+		if def_ptu and def_ptu.on_rightclick then
```
Comment (rubenwardy) [inline mods/default/nodes.lua]: shouldn't shift override the pointed at node's right click? To allow you to place anyway

Comment (paramat) [reply mods/default/nodes.lua]: Ok i'll add that.

**Comment by paramat** [comment]
Updated, now testing.

**Comment by paramat** [comment]
eb5a5b56e194bbaa432edea9697f1adb1d2398d9

**Pull request closed without merging**